polaris-service/src/test/java/org/apache/polaris/service/test/TestEnvironmentExtension.java

```diff
@@ -0,0 +1,115 @@
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+
+/**
+ * JUnit test extension that constructs a TestEnvironment. It tries to use environment variables and
+ * falls back to communicating with the local Dropwizard instance.
+ */
+public class TestEnvironmentExtension implements ParameterResolver {
+  /**
+   * Environment variable that specifies the Polaris base URL to use. If this is not set, falls back
+   * to the root path of your Dropwizard instance.
+   */
+  public static final String ENV_BASE_URL = "INTEGRATION_TEST_BASE_URL";
+
+  /**
+   * Environment variable that specifies the HTTP client factory that should construct test HTTP
+   * clients. This should be a class name. If this is not set, falls back to the client created by
+   * Dropwizard.
+   */
+  public static final String ENV_HTTP_CLIENT_FACTORY_IMPL =
+      "INTEGRATION_TEST_HTTP_CLIENT_FACTORY_IMPL";
+
+  private static TestEnvironment env;
+
+  public static synchronized TestEnvironment getEnv(ExtensionContext extensionContext)
+      throws IllegalAccessException {
+    if (env == null) {
+      Optional<String> baseUrl = Optional.ofNullable(System.getenv(ENV_BASE_URL));
+      DropwizardAppExtension dropwizardAppExtension = findDropwizardExtension(extensionContext);
+      if (dropwizardAppExtension == null && baseUrl.isEmpty()) {
+        throw new ParameterResolutionException(
+            "No test URL specified. Tried to default to Dropwizard but could not find DropwizardAppExtension.");
+      }
+
+      env =
+          new TestEnvironment(
+              getHttpClient(dropwizardAppExtension),
+              baseUrl.orElse(
+                  String.format("http://localhost:%d", dropwizardAppExtension.getLocalPort())));
+    }
+    return env;
+  }
+
+  @Override
+  public boolean supportsParameter(
+      ParameterContext parameterContext, ExtensionContext extensionContext)
+      throws ParameterResolutionException {
+    return parameterContext.getParameter().getType().equals(TestEnvironment.class);
+  }
+
+  @Override
+  public Object resolveParameter(
+      ParameterContext parameterContext, ExtensionContext extensionContext)
+      throws ParameterResolutionException {
+    try {
+      return getEnv(extensionContext);
+    } catch (IllegalAccessException e) {
+      throw new ParameterResolutionException(e.getMessage());
+    }
+  }
+
+  private static Client getHttpClient(DropwizardAppExtension dropwizardAppExtension) {
+    Optional<String> httpClientImpl =
+        Optional.ofNullable(System.getenv(ENV_HTTP_CLIENT_FACTORY_IMPL));
+
+    if (httpClientImpl.isEmpty()) {
+      if (dropwizardAppExtension == null) {
+        throw new ParameterResolutionException(
+            "Tried to default to Dropwizard client but could not find DropwizardAppExtension.");
+      }
+      return dropwizardAppExtension.client();
```

Comment:
Should this logic actually get encapsulated in (e.g.) `DropWizardTestHttpClientFactory`?

Reply:
Updated to TestEnvironmentResolver + DropwizardTestEnvironmentResolver